File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self) -> None:
        self.count = 0

    def __kthSmallestHelp(self, curNode: TreeNode, k:int) -> int:
        if not curNode:
            return 0
        
        # if curNode.left:
        leftResult = self.__kthSmallestHelp(curNode.left, k)
        if leftResult:
            return leftResult

        self.count += 1
        if self.count == k:
            return curNode.val

        # if curNode.right:
        return self.__kthSmallestHelp(curNode.right, k)          

    def kthSmallest(self, root: TreeNode, k: int) -> int:
        return self.__kthSmallestHelp(root, k)

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int cnt = 0;
        return findNode(root, k, &cnt);
    }

private:
    int findNode(TreeNode* node, int k, int* cnt) {
        if (node == nullptr) {
            return INT_MIN; // or any suitable default/error value
        }
        
        int left = findNode(node->left, k, cnt);
        if (left != INT_MIN) {
            return left;
        }

        (*cnt)++;
        if (*cnt == k) {
            return node->val;
        }

        return findNode(node->right, k, cnt);
    }
};